osu.Server.Spectator/Hubs/Multiplayer/ServerMultiplayerRoom.cs

```diff
@@ -29,22 +31,87 @@ public IMatchController Controller
         {
+            ServerMultiplayerRoom room = new ServerMultiplayerRoom(roomId, hub, dbFactory);
+
+            // TODO: this call should be transactional, and mark the room as managed by this server instance.
+            // This will allow for other instances to know not to reinitialise the room if the host arrives there.
+            // Alternatively, we can move lobby retrieval away from osu-web and not require this in the first place.
+            // Needs further discussion and consideration either way.
             using (var db = dbFactory.GetInstance())
             {
-                foreach (var item in await db.GetAllPlaylistItemsAsync(RoomID))
-                    Playlist.Add(item.ToMultiplayerPlaylistItem());
+                hub.Log(room, null, $"Retrieving room {roomId} from database");
+                var databaseRoom = await db.GetRealtimeRoomAsync(roomId);
+
+                if (databaseRoom == null)
+                    throw new InvalidOperationException("Specified match does not exist.");
+
+                if (databaseRoom.ends_at != null && databaseRoom.ends_at < DateTimeOffset.Now)
+                    throw new InvalidStateException("Match has already ended.");
+
+                room.ChannelID = databaseRoom.channel_id;
+                room.Settings = new MultiplayerRoomSettings
+                {
+                    Name = databaseRoom.name,
+                    Password = databaseRoom.password,
+                    MatchType = databaseRoom.type.ToMatchType(),
+                    QueueMode = databaseRoom.queue_mode.ToQueueMode(),
+                    AutoStartDuration = TimeSpan.FromSeconds(databaseRoom.auto_start_duration),
+                    AutoSkip = databaseRoom.auto_skip
+                };
+
+                foreach (var item in await db.GetAllPlaylistItemsAsync(roomId))
+                    room.Playlist.Add(item.ToMultiplayerPlaylistItem());
+
+                await room.ChangeMatchType(room.Settings.MatchType);
+
+                hub.Log(room, null, "Marking room active");
+                await db.MarkRoomActiveAsync(room);
+            }
+
+            return room;
+        }
+
+        /// <summary>
+        /// <see cref="InitialiseAsync">Initialises</see> a matchmaking room with the given eligible user IDs.
+        /// </summary>
+        /// <param name="roomId">The room identifier.</param>
+        /// <param name="hub">The multiplayer hub context.</param>
+        /// <param name="dbFactory">The database factory.</param>
+        /// <param name="eligibleUserIds">The users who are allowed to join the room.</param>
+        /// <exception cref="InvalidOperationException">If the room is not a matchmaking room in the database.</exception>
+        public static async Task<ServerMultiplayerRoom> InitialiseMatchmakingRoomAsync(long roomId, IMultiplayerHubContext hub, IDatabaseFactory dbFactory, int[] eligibleUserIds)
```

Comment:
Not 100% sure this is the best class to house this. Could it be more local to the usage?